llama-index-integrations/storage/chat_store/llama-index-storage-chat-store-upstash/llama_index/storage/chat_store/upstash/base.py

```diff
@@ -0,0 +1,210 @@
+from llama_index.core.llms import ChatMessage
+from llama_index.core.storage.chat_store.base import BaseChatStore
+
+import logging
+from typing import List, Optional, Any
+from upstash_redis import Redis
+from llama_index.core.bridge.pydantic import Field
+import json
+
+logger = logging.getLogger(__name__)
+
+
+# Convert a ChatMessage to a json object for Redis
+def _message_to_dict(message: ChatMessage) -> dict:
+    """
+    Convert a ChatMessage to a JSON-serializable dictionary.
+
+    Args:
+        message (ChatMessage): The ChatMessage object to convert.
+
+    Returns:
+        dict: A dictionary representation of the ChatMessage.
+    """
+    return message.dict()
+
+
+class UpstashChatStore(BaseChatStore):
+    """
+    Upstash chat store for storing and retrieving chat messages using Redis.
+
+    This class implements the BaseChatStore interface and provides methods
+    for managing chat messages in an Upstash Redis database.
+    """
+
+    redis_client: Any = Field(description="Redis client.")
+    ttl: Optional[int] = Field(default=None, description="Time to live in seconds.")
```

Comment:
This seems like just a wrapper on top of redis?

Any reason that the existing redis chat store wouldn't work? Or should we subclass the redis chat store?

The redis chat store also has async support which is not present here

Reply:
Hey @logan-markewich some serverless environments (like Deno, Cloudflare Workers etc) don't support TCP
That's why only Rest HTTP based SDKs work on those envs. Upstash Redis solves that problem.

Reply:
We'll also add async support

Reply:
Maybe I'm still confused though, the existing redis chat uses the same underlying redis client.

Why can't I configure the existing redis chat store to point to an upstash URL?

https://github.com/run-llama/llama_index/blob/a620a2661faabb49ba2f257bff7ae2ac04d0c12b/llama-index-integrations/storage/chat_store/llama-index-storage-chat-store-redis/llama_index/storage/chat_store/redis/base.py#L33

Reply:
So Upstash provides two ways to connect to databases:  TCP connection or via REST API. You can indeed use RedisChatStore to connect to Upstash, using TCP credentials. However, if you're on a serverless environment, TCP connection is not a feasible option, you should keep creating new connections since you can't use connection pooling etc.. This integration uses `upstash_redis`, and unlike the client RedisChatStore uses, `upstash_redis` client is HTTP based. Redis normally doesn't provide this option. So serverless people etc. can use this